managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -268,15 +277,53 @@ protected void recoverFromLedger(final ManagedCursorInfo info, final VoidCallbac
                 }
 
                 PositionImpl position = new PositionImpl(positionInfo);
+                PositionImpl lastIndividualAckedMessage = null;
                 if (positionInfo.getIndividualDeletedMessagesCount() > 0) {
                     recoverIndividualDeletedMessages(positionInfo.getIndividualDeletedMessagesList());
+                    // also recover readPosition to last acked message
+                    lastIndividualAckedMessage = new PositionImpl(
+                            positionInfo.getIndividualDeletedMessagesList().get(positionInfo.getIndividualDeletedMessagesCount() - 1).getUpperEndpoint());
                 }
                 recoveredCursor(position);
+                // if individualDeletedMessages are recovered then readPosition should be set after last Acked-Message
+                if (lastIndividualAckedMessage != null) {
+                    this.readPosition = lastIndividualAckedMessage;
+                }
                 callback.operationComplete();
             }, null);
         }, null);
     }
 
+    /**
+     * return unacked messages which are not part of individual deleted messages
+     * 
+     * @return
+     */
+    @Override
+    public Set<? extends Position> getNotDeletedMessages() {
```

Comment:
> When reading, the managed cursor is already skipping all the individually deleted messages.

Yes, that's correct, cursor reads messages in chunk of size 100 and then filter out already deleted messages and then keep going on. 
However, recently we have an issue for one of the subscription where there were large number of messages present between `markDeletePosition` and last acked message in `individuallyDeletedMessages` and out of which only few of them were unacked messages. So, cursor was keep reading from `markDeletePosition` in chunk of size 100 and filtering out most of the messages. so, `readPosition` was keep incrementing but consumer was not receiving the messages for long time.

Example: just added snapshot of `individuallyDeletedMessages` and it was large
```
: {
      "markDeletePosition" : "402212931:36356",
      "readPosition" : "403263473:272",
      "waitingReadOp" : false,
      "pendingReadOps" : 1,
      "messagesConsumedCounter" : -385138,
      "cursorLedger" : -1,
      "cursorLedgerLastEntry" : -1,
      "individuallyDeletedMessages" : "[(402212931:36357‥402258887:15270], (402258887:15271‥402312619:6246], (402312619:6247‥402343507:21490], .......
.......
.......
.......
(405361786:172156‥405361786:172253], (405361786:172265‥405361786:172272]]",
    
```